cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -159,14 +181,22 @@ public void updateAppSpec(String namespaceId, String appId, String versionId, Ap
     LOG.trace("App spec to be updated: id: {}: spec: {}", appId, GSON.toJson(spec));
     MDSKey key = new MDSKey.Builder().add(TYPE_APP_META, namespaceId, appId, versionId).build();
     ApplicationMeta existing = getFirst(key, ApplicationMeta.class);
+    ApplicationMeta updated;
+
+    // Check again without the version to account for old data format if might not have been upgraded yet
+    if (existing == null && (versionId.equals(ApplicationId.DEFAULT_VERSION))) {
+      key = new MDSKey.Builder().add(TYPE_APP_META, namespaceId, appId).build();
+      existing = getFirst(key, ApplicationMeta.class);
+    }
+
     if (existing == null) {
       String msg = String.format("No meta for namespace %s app %s exists", namespaceId, appId);
       LOG.error(msg);
       throw new IllegalArgumentException(msg);
     }
 
+    updated = ApplicationMeta.updateSpec(existing, spec);
     LOG.trace("Application exists in mds: id: {}, spec: {}", existing);
-    ApplicationMeta updated = ApplicationMeta.updateSpec(existing, spec);
     write(key, updated);
```

Comment:
is there a possible race condition? If the background upgrade thread upgrades this entry between the read in line 189 and this line, then the update will be written with the old (version-less) key, but the upgrade thread just created an outdated versioned entry and deleted the old one.